kimchi/src/circuits/polynomials/foreign_field_add/gadget.rs

```diff
@@ -0,0 +1,414 @@
+            .as_ref()
+            .ok_or(CircuitGateError::MissingLookupConstraintSystem(self.typ))?;
+
+        let lookup_env_data = set_up_lookup_env_data(
+            self.typ,
+            cs,
+            &witness,
+            &beta,
+            &gamma,
+            &lcs.configuration.lookup_info,
+        )
+        .map_err(|e| e)?;
+        let lookup_env = Some(LookupEnvironment {
+            aggreg: &lookup_env_data.aggreg8,
+            sorted: &lookup_env_data.sorted8,
+            selectors: &lcs.lookup_selectors,
+            table: &lookup_env_data.joint_lookup_table_d8,
+            runtime_selector: None,
+            runtime_table: None,
+        });
+
+        // Set up the environment
+        let env = {
+            Environment {
+                constants: expr::Constants {
+                    alpha: F::rand(rng),
+                    beta: F::rand(rng),
+                    gamma: F::rand(rng),
+                    joint_combiner: Some(F::rand(rng)),
+                    endo_coefficient: cs.endo,
+                    mds: vec![], // TODO: maybe cs.fr_sponge_params.mds.clone()
+                    foreign_field_modulus: vec![],
+                },
+                witness: &witness_evals.d8.this.w,
+                coefficient: &cs.coefficients8,
+                vanishes_on_last_4_rows: &cs.precomputations().vanishes_on_last_4_rows,
+                z: &witness_evals.d8.this.z,
+                l0_1: l0_1(cs.domain.d1),
+                domain: cs.domain,
+                index: index_evals,
+                lookup: lookup_env,
+            }
+        };
+
+        // Setup powers of alpha
+        let mut alphas = Alphas::<F>::default();
+        alphas.register(
+            ArgumentType::Gate(self.typ),
+            circuit_gate_constraint_count::<F>(self.typ),
+        );
+
+        // Get constraints for this circuit gate
+        let constraints = circuit_gate_constraints(self.typ, &alphas);
+
+        // Verify it against the environment
+        if constraints
+            .evaluations(&env)
+            .interpolate()
+            .divide_by_vanishing_poly(cs.domain.d1)
+            .unwrap()
+            .1
+            .is_zero()
+        {
+            Ok(())
+        } else {
+            Err(CircuitGateError::InvalidConstraint(self.typ))
+        }
+    }
+}
+
+// Connect the pair of cells in different wires vectors, specified by the
+// wires1, col1 and wires2, cell2 parameters
+//
+// cell1 --> cell2 && cell2 --> cell1
+//
+// Cell format is: (row, col)
+//
+// Note: This function assumes that the targeted cells are freshly instantiated
+//       with self-connections.  If the two cells are transitively already part
+//       of the same permutation then this would split it.
+fn connect_cell_pair(
+    wires1: &mut GateWires,
+    col1: usize,
+    wires2: &mut [GateWires],
+    cell2: (usize, usize),
+) {
+    std::mem::swap(&mut wires1[col1], &mut wires2[cell2.0][cell2.1]);
+}
```

Comment:
Oh you're right... I think it was a private function in range_check and I just copied it again. In any case, this way of defining the wirings, and also filling in the witness with the `WitnessCell` pattern is something that, if we wanted to pursue, should be written somewhere more general like `gate.rs` as you're suggesting. It would definitely be moved, but I was waiting for either FFAdd or FFMul to be merged to shape this common format (so far, it is only being applied in FFadd, FFmul and RangeCheck)